src/GitHub.App/Services/TeamExplorerContext.cs

```diff
@@ -43,28 +47,56 @@ public class TeamExplorerContext : ITeamExplorerContext
         ILocalRepositoryModel repositoryModel;
 
         [ImportingConstructor]
+        TeamExplorerContext(
+            IVSGitExt gitExt,
+            [Import(typeof(SVsServiceProvider))] IServiceProvider sp,
+            IPullRequestService pullRequestService) : this(
+                gitExt,
+                new AsyncLazy<DTE>(async () =>
+                {
+                    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    return (DTE)sp.GetService(typeof(DTE));
+                }),
+                pullRequestService,
+                ThreadHelper.JoinableTaskContext)
+        {
+        }
+
         public TeamExplorerContext(
-            IGitHubServiceProvider serviceProvider,
             IVSGitExt gitExt,
-            IPullRequestService pullRequestService)
+            AsyncLazy<DTE> dteAsync,
+            IPullRequestService pullRequestService,
+            JoinableTaskContext joinableTaskContext)
         {
+            JoinableTaskCollection = joinableTaskContext.CreateCollection();
+            JoinableTaskCollection.DisplayName = nameof(TeamExplorerContext);
+            JoinableTaskFactory = joinableTaskContext.CreateFactory(JoinableTaskCollection);
+            
             this.gitExt = gitExt;
+            this.dteAsync = dteAsync;
             this.pullRequestService = pullRequestService;
 
-            // This is a standard service which should always be available.
-            dte = serviceProvider.GetService<DTE>();
-
-            Refresh();
+            StartRefresh();
             gitExt.ActiveRepositoriesChanged += Refresh;
         }
 
-        async void Refresh()
+        void StartRefresh() => JoinableTaskFactory.Context.Factory.RunAsync(QueueRefreshAsync).Task.FileAndForget(log);
```

Comment:
@AArnott I want to ensure that there are no overlapping calls to `RefreshAsync`. This seems to work, but I worry about other people looking at the code. Is there a more obvious way to do this?

I've just reminded myself of the following:
> Rule #3. If ever awaiting work that was started earlier, that work must be joined.

I'll see how it might look using `Join`.

Reply:
This is actually my own helper method to file exceptions in our own logs:
```cs
        public static void FileAndForget(this Task task, ILogger log)
        {
            task.ContinueWith(t =>
            {
                if (t.IsFaulted)
                {
                    log.Error(t.Exception, nameof(FileAndForget));
                }
            });
        }
```

Maybe I should call it `LogAndForget` to disambiguate?